File: Car.js
function dist(x1, y1, x2, y2) {
    var a = x1 - x2;
    var b = y1 - y2;
    return Math.sqrt( a*a + b*b );
}

class Car {
    constructor (w, h, len, dna) {
        this.w = w;
        this.h = h;
        this.dir = {x: 0, y: 1};
        this.drift = 20;
        this.start();
        this.actualizeWalls();
        this.deadAt = 400;
        this.checkpoints = [];
        if (dna) this.dna = dna;
        else {
            this.dna = new DNA(len);
        }
    }

    start() {
        this.x = 158;
        this.y = 239;
        this.rotation = -1.56;
        this.dir = {x: 1, y: 0};
        this.v = 0;
    }

    actualizeWalls() {
        let x1 = (this.x - this.w/2);
        let x4 = x1;
        let x2 = (this.x + this.w/2);
        let x3 = x2;
        
        let y1 = (this.y - this.h/2 + this.drift);
        let y2 = y1;
        let y3 = (this.y + this.h/2 + this.drift);
        let y4 = y3;

        let r = dist(x1, y1, this.x, this.y);
        x1 = this.x - r * Math.cos(this.rotation);
        y1 = this.y - r * Math.sin(this.rotation);
        x2 = this.x + r * Math.cos(this.rotation);
        y2 = this.y + r * Math.sin(this.rotation);
        
        let angle = Math.atan(this.h/(this.w/2));
        
        r = dist(x3, y3, this.x, this.y);
        x3 = this.x + r * Math.cos(this.rotation + angle);
        y3 = this.y + r * Math.sin(this.rotation + angle);
        x4 = this.x - r * Math.cos(this.rotation - angle);
        y4 = this.y - r * Math.sin(this.rotation - angle);
        
        this.walls = [
            new Wall(x1, y1, x2, y2), new Wall(x2, y2, x3, y3), 
            new Wall(x3, y3, x4, y4), new Wall(x4, y4, x1, y1)
        ];
    }

    show(color) {
        d.translate(this.x, this.y);
        d.rotate(this.rotation);
        
        // Cotxe
        d.rectangle(0, + this.drift, this.w, this.h, {color: color, alpha: 0.7});
        
        // --------------------------- Posible decorations ---------------------------
        // Llums
        // d.rectangle(0, drift - this.h/2, this.w, 5, {color: "yellow"});
        // Rodes
        // d.rectangle(-this.w/2 - 3, drift - this.h/2 + 6, 6, 8, {color: "black"});
        // d.rectangle(-this.w/2 - 3, drift + this.h/2 - 6, 6, 8, {color: "black"});
        // d.rectangle(this.w/2 + 3, drift - this.h/2 + 6, 6, 8, {color: "black"});
        // d.rectangle(this.w/2 + 3, drift + this.h/2 - 6, 6, 8, {color: "black"});
        // --------------------------- Posible decorations ---------------------------

        d.rotate(-this.rotation);
        d.translate(-this.x, -this.y);
    }

    move() {
        this.x += this.dir.x * this.v;
        this.y += this.dir.y * this.v;
        this.actualizeWalls();
    }

    action(a) {
        var vel = 1;
        var rot = 0.13;
        
        if (a == 0) this.forward(vel);
        else if (a == 1) this.back(vel);
        else if (a == 2) this.left(rot);
        else if (a == 3) this.right(rot);
        else if (a == 4) {
            // Nothing
        }
    }

    checkVel() {
        var a = 0.5;
        if (this.v > 0) {
            if (this.v >= a) this.v = this.v - a;
            if (this.v < a) this.v = 0; 
        }
        else if (this.v < 0) {
            if (this.v <= -a) this.v = this.v + a;
            if (this.v > -a) this.v = 0; 
        }
    }

    collides(w) {
        for (let i = 0; i < 4; ++i) {
            var aux = this.walls[i]; 
            let P = aux.intersection(w);
            if (P != undefined) {
                return true;
            }
        }
        return false;
    }

    forward(vel) {
        this.v -= vel;
    }

    back(vel) {
        this.v += vel;
    }

    left(rot) {
        const halfPI = Math.PI/2;
        this.rotation -= rot;
        this.dir.x = Math.cos(halfPI + this.rotation);
        this.dir.y = Math.sin(halfPI + this.rotation);
    }

    right(rot) {
        const halfPI = Math.PI/2;
        this.rotation += rot;
        this.dir.x = Math.cos(halfPI + this.rotation);
        this.dir.y = Math.sin(halfPI + this.rotation);
    }

    fitness() {
        let c = this.checkpoints;
        if (c[0] == undefined || c[0][0] != 0) return 0;
        let fitness = 0;

        for (let i = 0; i < c.length; ++i) {
            let wall = c[i];
            let t = wall[1];
            fitness += ((wall[0] + 1)*10)/(t);
        }
        return fitness;
    }

    sex(partner) {
        let dnaX = this.dna; 
        let dnaY = partner.dna; 
        let childDNA = dnaX.crossover(dnaY);
        return new Car(this.h, this.w, this.len, childDNA);
    }

    print() {
        let gens = this.dna.gens;
        let s = "[";
        for (let i = 0; i < gens.length; ++i) {
            s += gens[i] + ',';
        }
        s += ']';
        console.log(s);
    }

    copy() {
        return new Car(this.w, this.h, this.len, this.dna.copy());
    }

    mutate(r) {
        for (let i = 0; i < this.dna.gens.length; ++i) {
            if (Math.random() < r) {
                this.dna.gens[i] = Math.floor(Math.random()*4);
            }
        }
    }
}

File: drift-v2/car.py
import math
from wall import Wall 

def dist(x1, y1, x2, y2):
    a = x1 - x2
    b = y1 - y2
    return math.sqrt(a * a + b * b)

class Car ():
    def __init__ (self, w, h, route_walls, brain):
        self.w = w
        self.h = h
        self.drift = 20
        self.sense_walls = []
        self.sense_dists = []
        self.sense_dots = []
        self.vision_len = 400
        self.brain = brain
        self.route = route_walls
        self.restart()
        self.actualize_walls()

    def restart(self):
        self.checkpoints = []
        self.x = 158
        self.y = 239
        self.rotation = -1.56
        self.dir = (1, 0) # x, y
        self.v = 0
        self.dead_at = 400

    def actualize_route_walls(self):

        self.sense_walls = [0] * 8
        x1 = self.x
        y1 = self.y
        pi4 = math.pi/4

        for i in range(8):
            angle = self.rotation + i *pi4
            x2 = x1 + self.vision_len * math.cos(angle)
            y2 = y1 + self.vision_len * math.sin(angle)
            self.sense_walls[i] = Wall(x1, y1, x2, y2)

        for i in range(8):
            min_dot = None
            min_dist = math.inf
            
            for j in range(len(self.route)):
                collision = self.sense_walls[i].intersection(self.route[j])
                if collision:
                    distance = dist(collision[0], collision[1], self.x, self.y)
                    if distance < min_dist:
                        min_dist = distance
                        min_dot = collision
            
            self.sense_dots[i] = min_dot
            
            if min_dist == None or min_dist > self.vision_len:
                min_dist = self.vision_len
            
            self.sense_dists[i] = min_dist / self.vision_len
    
    def actualize_walls(self):
        self.actualizeRouteWalls()

        # Actualització parets que formen el cotxe
        x1 = (self.x - self.w/2)
        x4 = x1
        x2 = (self.x + self.w/2)
        x3 = x2
        
        y1 = (self.y - self.h/2 + self.drift)
        y2 = y1
        y3 = (self.y + self.h/2 + self.drift)
        y4 = y3

        r = dist(x1, y1, self.x, self.y)
        x1 = self.x - r * math.cos(self.rotation)
        y1 = self.y - r * math.sin(self.rotation)
        x2 = self.x + r * math.cos(self.rotation)
        y2 = self.y + r * math.sin(self.rotation)
        
        angle = math.atan(self.h/(self.w/2))
        
        r = dist(x3, y3, self.x, self.y)
        x3 = self.x + r * math.cos(self.rotation + angle)
        y3 = self.y + r * math.sin(self.rotation + angle)
        x4 = self.x - r * math.cos(self.rotation - angle)
        y4 = self.y - r * math.sin(self.rotation - angle)
        
        self.walls = [
            Wall(x1, y1, x2, y2), Wall(x2, y2, x3, y3), 
            Wall(x3, y3, x4, y4), Wall(x4, y4, x1, y1)
        ]

    def move(self):
        self.x += self.dir[0] * self.v
        self.y += self.dir[1] * self.v
        self.check_vel()
        self.actualize_walls()

    def think(self):
        vel = 1
        rot = 0.13
        th = 0.8

        res = self.brain.out(self.sense_dists) # TODO: pensar bé això
        if res[0] > th: self.forward(vel)
        else if res[1] > th: self.back(vel)
        if res[2] > th: self.left(rot)
        else if res[3] > th: self.right(rot)

    def action(self, a):
        vel = 1
        rot = 0.13

        if a == 0: self.forward(vel)
        elif a == 1: self.back(vel)
        elif a == 2: self.left(rot)
        elif a == 3: self.right(rot)
        elif a == 4: pass

    def check_vel(self):
        a = 0.5
        if self.v > 0:
            if self.v >= a: self.v = self.v - a
            if self.v < a: self.v = 0 
        
        else if self.v < 0:
            if self.v <= -a: self.v = self.v + a
            if self.v > -a: self.v = 0 
        
    def collides(self, w):
        for i in range(4):
            P = self.walls[i].intersection(w)
            if (P) return True
        return False
    
    def forward(self, vel):
        self.v -= vel
    
    def back(self, vel):
        self.v += vel
    
    def left(self, rot):
        halfpi = math.pi/2
        self.rotation += rot
        self.dir[0] = math.cos(halfpi + self.rotation)
        self.dir[1] = math.sin(halfpi + self.rotation)

    def right(self, rot):
        halfpi = math.pi/2
        self.rotation += rot
        self.dir[0] = math.cos(halfpi + self.rotation)
        self.dir[1] = math.sin(halfpi + self.rotation)

    def fitness (self):
        c = self.checkpoints
        if len(c) == 0: return 0
        if c[0][0] != 0: return 0

        fitness = 0
        for i in range(len(c)):
            t = 0
            if i > 0: t = c[i][1] - c[i - 1][1]
            else: t = c[i][1]
            nwalls = c[i][0]
            fitness += ((nwalls + 1) * 10) + (10 * nwalls/t)
        
        return fitness

    def crossover(self, other):
        other_fit = other.fitness()
        best = other_fit > this.fitness()
        new_brain = self.brain.crossover(other.brain, best)
        return new_brain